pg_search/src/postgres/customscan/hook.rs

```diff
@@ -211,3 +216,433 @@ pub extern "C-unwind" fn paradedb_upper_paths_callback<CS>(
+        let rtable = PgList::<pg_sys::RangeTblEntry>::from_pg((*parse).rtable);
+        for (idx, rte) in rtable.iter_ptr().enumerate() {
+            if (*rte).rtekind == pg_sys::RTEKind::RTE_SUBQUERY
+                && !(*rte).subquery.is_null()
+                && query_has_window_aggregates((*rte).subquery)
+            {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+/// Check if the query contains the @@@ search operator
+/// This indicates that our custom scans will likely handle this query
+/// Uses expression_tree_walker via expr_contains_any_operator for complete traversal
+unsafe fn query_has_search_operator(parse: *mut pg_sys::Query) -> bool {
+    let searchqueryinput_opno = anyelement_query_input_opoid();
+    let text_opno = anyelement_text_opoid();
+    let target_ops = [searchqueryinput_opno, text_opno];
+
+    // Check WHERE clause (jointree->quals)
+    if !(*parse).jointree.is_null() {
+        let jointree = (*parse).jointree;
+        if !(*jointree).quals.is_null()
+            && expr_contains_any_operator((*jointree).quals, &target_ops)
+        {
+            return true;
+        }
+    }
+
+    // Check HAVING clause
+    if !(*parse).havingQual.is_null()
+        && expr_contains_any_operator((*parse).havingQual, &target_ops)
+    {
+        return true;
+    }
+
+    // Check target list (for search operators in SELECT expressions, aggregates, window functions)
+    if !(*parse).targetList.is_null() {
+        let target_list = PgList::<pg_sys::TargetEntry>::from_pg((*parse).targetList);
+        for te in target_list.iter_ptr() {
+            if !(*te).expr.is_null()
+                && expr_contains_any_operator((*te).expr as *mut pg_sys::Node, &target_ops)
+            {
+                return true;
+            }
+        }
+    }
+
+    // Check subqueries recursively
+    if !(*parse).rtable.is_null() {
+        let rtable = PgList::<pg_sys::RangeTblEntry>::from_pg((*parse).rtable);
+        for rte in rtable.iter_ptr() {
+            if (*rte).rtekind == pg_sys::RTEKind::RTE_SUBQUERY
+                && !(*rte).subquery.is_null()
+                && query_has_search_operator((*rte).subquery)
+            {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+/// Check if the query is a TopN query (has ORDER BY and LIMIT)
+unsafe fn query_is_topn(parse: *mut pg_sys::Query) -> bool {
+    if parse.is_null() {
+        return false;
+    }
+
+    // Must have both ORDER BY (sortClause) and LIMIT (limitCount)
+    let has_order_by = !(*parse).sortClause.is_null();
+    let has_limit = !(*parse).limitCount.is_null();
+
+    has_order_by && has_limit
+}
+
+/// Check if the query contains paradedb.agg() in any context (window function or aggregate)
+/// If it does, we MUST handle it (even without @@@ operator)
+/// Uses expression_tree_walker for complete traversal
+unsafe fn query_has_paradedb_agg(parse: *mut pg_sys::Query) -> bool {
```

Comment:
Ditto duped with `maybe_needs_const_projections`. Please add comments to both of them as a reminder to keep them in sync.